firedrake/extrusion_numbering.pyx

```diff
@@ -0,0 +1,1035 @@
+        in the masking arrays for the start of each column.
+    """
+    cdef:
+        PETSc.Section section, cell_numbering, cell_section
+        DMLabel lbl
+        PETSc.DM dm
+        PetscInt pStart, pEnd, fStart, fEnd
+        PetscInt supportSize
+        const PetscInt *support = NULL
+        numpy.ndarray[PetscInt, ndim=1, mode="c"] renumbering
+        numpy.ndarray[PetscInt, ndim=2, mode="c"] layer_extents
+        numpy.ndarray[numpy.int64_t, ndim=1, mode="c"] top_mask, bottom_mask, cell_bottom_mask, cell_top_mask
+        PetscInt p, nfacet, facet, point, ent_bottom, ent_top, cell_bottom, cell_offset, facet_offset
+        PetscInt i, j, c, cell
+        numpy.int64_t nentities
+        PetscBool flg
+        
+    assert label in {"interior_facets", "exterior_facets"}
+
+    label = label.encode()
+    dm = mesh._plex
+    pStart, pEnd = dm.getChart()
+    fStart, fEnd = dm.getHeightStratum(1)
+
+    cell_numbering = mesh._cell_numbering
+    renumbering = mesh._plex_renumbering.indices
+
+
+    nfacet = layers.shape[0]
+    section = PETSc.Section().create(comm=PETSc.COMM_SELF)
+    section.setChart(0, layers.shape[0])
+    for p in range(nfacet):
+        CHKERR(PetscSectionSetDof(section.sec, p, layers[p, 1] - layers[p, 0] - 1))
+    section.setUp()
+    top_mask = numpy.zeros(section.getStorageSize(), dtype=numpy.int64)
+    bottom_mask = numpy.zeros_like(top_mask)
+
+    fiat_cell = as_fiat_cell(mesh.ufl_cell())
+    # Number of topological entities on the cell (excluding cell itself)
+    nentities = sum(map(len, fiat_cell.get_topology().values())) - 1
+    assert nentities < 31, "Need fewer than 31 entities"
+
+    cell_section, cell_bottom_mask, cell_top_mask = mesh.cell_set.masks
+    facet = 0
+    layer_extents = mesh.layer_extents
+    CHKERR(DMGetLabel(dm.dm, <const char *>label, &lbl))
+    CHKERR(DMLabelCreateIndex(lbl, fStart, fEnd))
+    for p in range(pStart, pEnd):
+        point = renumbering[p]
+        if point < fStart or point >= fEnd:
+            continue
+        CHKERR(DMLabelHasPoint(lbl, point, &flg))
+        if not flg:
+            continue
+        ent_bottom = layer_extents[point, 2]
+        ent_top = layer_extents[point, 3]
+        CHKERR(DMPlexGetSupportSize(dm.dm, point, &supportSize))
+        CHKERR(DMPlexGetSupport(dm.dm, point, &support))
+        for j in range(supportSize):
+            c = support[j]
+            CHKERR(PetscSectionGetOffset(cell_numbering.sec, c, &cell))
+            cell_bottom = layer_extents[c, 0]
+            CHKERR(PetscSectionGetOffset(cell_section.sec, cell, &cell_offset))
+            CHKERR(PetscSectionGetOffset(section.sec, facet, &facet_offset))
+            for i in range(ent_bottom, ent_top - 1):
+                top_mask[facet_offset + i - ent_bottom] |= cell_top_mask[cell_offset + i - cell_bottom] << (nentities*j)
+                bottom_mask[facet_offset + i - ent_bottom] |= cell_bottom_mask[cell_offset + i - cell_bottom] << (nentities*j)
```

Comment:
So the facet masks are just the concatenation of the cell masks in the support of the facet.  Note that I only need to maintain facet masks for "iteration" facets, because I've guaranteed already that I will never assemble over the exposed facets, and so will never have to do any of these horrible tricks.

The first cell's mask is in the bottom `nentities` bits, the second cell's in the top `nentities` bits.  So in fact for most cell types we could get away with 32bits, but I don't care enough.